src/platform/ESP32/ESP32TCPEndPointImplLwIP.cpp

```diff
@@ -0,0 +1,1036 @@
+        if (err == CHIP_NO_ERROR)
+            OnConnectionReceived(this, conEP, peerAddr, peerPort);
+
+        // Otherwise clean up and call the app's error callback.
+        else if (OnAcceptError != NULL)
+            OnAcceptError(this, err);
+    }
+    else
+        err = CHIP_ERROR_INCORRECT_STATE;
+
+    // If something failed above, abort and free the connection end point.
+    if (err != CHIP_NO_ERROR)
+        conEP->Free();
+}
+
+void TCPEndPointImplLwIP::HandleError(CHIP_ERROR err)
+{
+    if (mState == State::kListening)
+    {
+        if (OnAcceptError != NULL)
+            OnAcceptError(this, err);
+    }
+    else
+        DoClose(err, false);
+}
+
+err_t TCPEndPointImplLwIP::LwIPHandleConnectComplete(void * arg, struct tcp_pcb * tpcb, err_t lwipErr)
+{
+    err_t res = ERR_OK;
+
+    if (arg != NULL)
+    {
+        TCPEndPointImplLwIP * ep = static_cast<TCPEndPointImplLwIP *>(arg);
+
+        if (lwipErr == ERR_OK)
+        {
+            // Setup LwIP callback functions for data transmission.
+            tcp_recv(ep->mTCP, LwIPHandleDataReceived);
+            tcp_sent(ep->mTCP, LwIPHandleDataSent);
+        }
+
+        // Post callback to HandleConnectComplete.
+        ep->Retain();
+        CHIP_ERROR err = ep->GetSystemLayer().ScheduleLambda([ep, conErr = System::MapErrorLwIP(lwipErr)] {
+            ep->HandleConnectComplete(conErr);
+            ep->Release();
+        });
+        if (err != CHIP_NO_ERROR)
+        {
+            ep->Release();
+            res = ERR_ABRT;
+        }
+    }
+    else
+        res = ERR_ABRT;
+
+    if (res != ERR_OK)
+        tcp_abort(tpcb);
+
+    return res;
+}
+
+err_t TCPEndPointImplLwIP::LwIPHandleIncomingConnection(void * arg, struct tcp_pcb * tpcb, err_t lwipErr)
+{
+    CHIP_ERROR err = chip::System::MapErrorLwIP(lwipErr);
+
+    if (arg != NULL)
+    {
+        TCPEndPointImplLwIP * listenEP = static_cast<TCPEndPointImplLwIP *>(arg);
+        TCPEndPointImplLwIP * conEP    = NULL;
+        System::Layer & lSystemLayer   = listenEP->GetSystemLayer();
+
+        // Tell LwIP we've accepted the connection so it can decrement the listen PCB's pending_accepts counter.
+        tcp_accepted(listenEP->mTCP);
+
+        // If we did in fact receive a connection, rather than an error, attempt to allocate an end point object.
+        //
+        // NOTE: Although most of the LwIP callbacks defer the real work to happen on the endpoint's thread
+        // (by posting events to the thread's event queue) we can't do that here because as soon as this
+        // function returns, LwIP is free to begin calling callbacks on the new PCB. For that to work we need
+        // to have an end point associated with the PCB.
+        //
+        if (err == CHIP_NO_ERROR)
+        {
+            TCPEndPoint * connectEndPoint = nullptr;
+            {
+                //lock CHIP stack
+                chip::DeviceLayer::StackLock lock;
```

Comment:
I had to copy TCPEndPointImplLwIP.cpp from src/inet to src/platform/ESP32 since it is required to use chip stack lock here. It is checked here src/inet/InetLayer.h:82 (https://github.com/project-chip/connectedhomeip/blob/1d7477934bbfaaaba3556d60a617c071830acf42/src/inet/InetLayer.h#L82)

Reply:
What are the differences between this file and the default inet/TCPEndPointImplLwIP.cpp? We shouldn't have to make a copy for ESP32 platform but only have one for all LwIP based platforms.
The locking should happen higher up at the application layer so lower down we have the assert checks to ensure that the lock is held.

Reply:
Differences:
1) Added LOCK_TCPIP_CORE() and UNLOCK_TCPIP_CORE() in TCPEndPointImplLwIP::ListenImpl due to `assert failed: tcp_listen_with_backlog /IDF/components/lwip/lwip/src/core/tcp.c:834 (Required to lock TCPIP core functionality!)`

2) Added lock in TCPEndPointImplLwIP::LwIPHandleIncomingConnection as mentioned above due to 
```
E (17377) chip[DL]: Chip stack locking error at 'src/inet/InetLayer.h:82'. Code is unsafe/racy
E (17377) chip[-]: chipDie chipDie chipDie
```
I had to make a copy because it was not allowed to include <platform/PlatformManager.h>
```
ERROR at //third_party/connectedhomeip/src/inet/TCPEndPointImplLwIP.cpp:32:11: Include not allowed.
#include <platform/PlatformManager.h>
          ^-------------------------
It is not in any dependency of
  //third_party/connectedhomeip/src/inet:inet
The include file is in the target(s):
  //third_party/connectedhomeip/src/platform:platform
which should somehow be reachable.
```

Reply:
The TCPEndPointImplLwIP is a common piece of code for all platforms using LwIP. It is not a virtual interface that needs to be implemented by individual platforms.
For 1), The LWIP core lock can be added in TCPEndPointImplLwIP.cpp instead of this copy in ESP32, no?
For 2) have you tried acquiring the CHIP Stack lock higher up, maybe around InitChipStack()?